module/state_synchronization/indexer/indexer_core.go

```diff
@@ -19,23 +22,29 @@ type IndexerCore struct {
 	registers storage.RegisterIndex
 	headers   storage.Headers
 	events    storage.Events
+	results   storage.LightTransactionResults
 	log       zerolog.Logger
+	db        *badger.DB
```

Comment:
I wish we could avoid passing an instance to the specific badger db, but I guess batch requires it? I don't like having a dependency on a specific implementation. I know this may call for a refactor, but my take on this would be to have an interface for batch (this is now even more true with implementing pebble since the same pattern will have to be established) and then have methods optionally supporting batch. Something like pseudo code:

```go
batch := batcher.New()

s.events.Store(blockID, []flow.EventsList{result.AllEvents()}, storage.WithBatch(batch))
s.serviceEvents.BatchStore(blockID, result.AllServiceEvents(), storage.WithBatch(batch))

batch.Flush()
```

And then batcher instance is passed in as:

```go
indexer.New(log, badger.NewBatcher(), registers, headers, events, results)
```

I feel this would be a nice pattern because it will also allow the same for Pebble, but now the Batch type assumes only badger and is very implementation-specific. Just a thought.

Reply:
Agree

Reply:
I think this is a good idea. There are other instances in the code where we do this so I'm reluctant to fold a larger refactor into this PR.

How about updating `storage.NewBatch` to take an interface instead of the whole `*badger.DB` instance? then all this needs is that interface?

Reply:
added here: [f8c0c34](https://github.com/onflow/flow-go/pull/4772/commits/f8c0c34974c78f3f0032795ce7f7ec2b6a116309)